src/main/java/org/opensearch/neuralsearch/processor/chunker/DelimiterChunker.java

```diff
@@ -62,23 +75,59 @@ public List<String> chunk(final String content, final Map<String, Object> runtim
         int runtimeMaxChunkLimit = parseInteger(runtimeParameters, MAX_CHUNK_LIMIT_FIELD);
         int chunkStringCount = parseInteger(runtimeParameters, CHUNK_STRING_COUNT_FIELD);
 
+        // Skip splitting to save memory if content is short and has no delimiter.
+        if (this.chunkSize > 0 && content.length() <= this.chunkSize && !content.contains(delimiter)) {
+            return Collections.singletonList(content);
+        }
+
         List<String> chunkResult = new ArrayList<>();
-        int start = 0, end;
-        int nextDelimiterPosition = content.indexOf(delimiter);
+        int start = 0;
 
-        while (nextDelimiterPosition != -1) {
+        while (start < content.length()) {
             if (Chunker.checkRunTimeMaxChunkLimit(chunkResult.size(), runtimeMaxChunkLimit, chunkStringCount)) {
+                chunkResult.add(content.substring(start));
                 break;
             }
-            end = nextDelimiterPosition + delimiter.length();
-            chunkResult.add(content.substring(start, end));
-            start = end;
-            nextDelimiterPosition = content.indexOf(delimiter, start);
-        }
 
-        // add the rest content into the chunk result
-        if (start < content.length()) {
-            chunkResult.add(content.substring(start));
+            int nextDelimiterPosition = content.indexOf(delimiter, start);
+            int end = (nextDelimiterPosition == -1) ? content.length() : nextDelimiterPosition + delimiter.length();
+
+            String segment = content.substring(start, end);
+
+            // 'segment' represents a portion of content sliced by the delimiter.
+            // If segment is too large, split it further by chunkSize
+            if (this.chunkSize > 0 && segment.length() > this.chunkSize) {
+                List<String> splitChunks = splitByLength(segment, this.chunkSize);
+                int segmentOffsetWithinContent = 0;
+
+                for (int i = 0; i < splitChunks.size(); i++) {
+                    String chunk = splitChunks.get(i);
+
+                    // If adding this chunk exceeds the max chunk limit, merge all remaining parts into the last chunk
+                    if (Chunker.checkRunTimeMaxChunkLimit(chunkResult.size(), runtimeMaxChunkLimit, chunkStringCount)) {
+                        StringBuilder lastChunkBuilder = new StringBuilder();
+
+                        for (int j = i; j < splitChunks.size(); j++) {
+                            lastChunkBuilder.append(splitChunks.get(j));
+                        }
+
+                        int remainingContentStartIndex = start + segmentOffsetWithinContent;
```

Comment:
segmentOffsetWithinContent is updated when chunk is added to chunkResult in line 123. 
Shouldn't we simply do `chunkResult.add(content.substring(start + segmentOffsetWithinContent))` at line 110?